File: src/logic/CallbackListener.java
package logic;

import gui.Chat;
import gui.ChatMessageBlueprint;
import gui.MainWindow;
import org.jetbrains.annotations.NotNull;
import protocol.ProtocolCallback;

import java.net.DatagramPacket;
import java.net.InetAddress;
import java.util.ArrayList;

public class CallbackListener implements ProtocolCallback {

    private MainWindow mainWindow;
    private Contacts contacts;
    private Groups groups;

    public int currentChatId = -1;

    public CallbackListener(MainWindow mainWindow, Contacts contacts, Groups groups) {
        this.mainWindow = mainWindow;

        contacts.setParent(this);
        groups.setParent(this);

        this.contacts = contacts;
        this.groups = groups;
    }

    @Override
    public void hello(@NotNull DatagramPacket packet, @NotNull String username) {
        Contact contact = contacts.createContact(packet.getAddress(), username, UserColors.getRandomColor());
        // debug
        if (contact == null) throw new RuntimeException();
        mainWindow.addContact(contact.getUsername(), "is now online", contact.getColor(), contact.getId());
        mainWindow.addNewChatById(contact.getId());

        if (!contact.getMessages().isEmpty())
            mainWindow.addMessages((contact.getMessages().toArray(new ChatMessageBlueprint[contact.getMessages().size()])), contact.getId());
    }

    @Override
    public void world(@NotNull DatagramPacket packet, @NotNull String username) {
        Contact contact = contacts.createContact(packet.getAddress(), username,
                UserColors.getRandomColor());
        // debug
        if (contact == null) throw new RuntimeException();
        mainWindow.addNewChatById(contact.getId());
        mainWindow.addContact(contact.getUsername(), "is now online", contact.getColor(), contact.getId());

        if (!contact.getMessages().isEmpty())
            mainWindow.addMessages((contact.getMessages().toArray(new ChatMessageBlueprint[contact.getMessages().size()])), contact.getId());
    }

    @Override
    public void goodbye(@NotNull DatagramPacket packet) {
        Contact contact = contacts.getByIP(packet.getAddress());
        // debug
        if (contact == null) throw new RuntimeException();
        mainWindow.addMessage(new ChatMessageBlueprint(Chat.chatMessageType.INFO, "ignored parameter",
                contact.getUsername() + " has left the room.", "",
                contact.getColor()), contact.getId());
        mainWindow.setLastMessageText("is now offline", contact.getId());
    }

    @Override
    public void typing(@NotNull DatagramPacket packet, boolean typing) {
        if (typing) {
            //TODO wtf
        } else {
            mainWindow.setContactWriting(contacts.getByIP(packet.getAddress()).getId());
        }
    }

    @Override
    public void message(@NotNull DatagramPacket packet, @NotNull String message) {
        Contact contact = contacts.getByIP(packet.getAddress());
        if (contact == null) {
            System.out.println("Message from unknown contact received: " + packet.getAddress());
        } else {
            ChatMessageBlueprint chatMessageBlueprint = new ChatMessageBlueprint(Chat.chatMessageType.FROM,
                    contact.getUsername(),
                    message, "", contact.getColor());

            mainWindow.addMessage(chatMessageBlueprint, contact.getId());
            mainWindow.setLastMessageText(message, contact.getId());

            contact.getMessages().add(chatMessageBlueprint);
        }
    }

    @Override
    public boolean existsGroupWithId(@NotNull DatagramPacket packet, int id) {
        // TODO: 19.12.2017 most likely should also return false, when a contact already uses this id
        return (groups.getByProtocolID(id) != null);
    }

    @Override
    public void createGroup(@NotNull DatagramPacket packet, int id, @NotNull InetAddress[] members) {
        ArrayList<Contact> membersAsContacts = new ArrayList<>();
        for (InetAddress memberAddress : members) {
            Contact contact = contacts.getByIP(memberAddress);

            // This should NEVER be true
            if (contact == null)
                throw new RuntimeException();

            membersAsContacts.add(contact);
        }
        groups.createGroup(id, membersAsContacts);
    }

    @Override
    public void denyGroup(@NotNull DatagramPacket packet) {
        // TODO: 15.12.2017 No clue what this means, or which steps should be taken.
    }

    @Override
    public void groupMessage(@NotNull DatagramPacket packet, int groupId, @NotNull String message) {
        // TODO: 19.12.2017 see todo
        int correspondingChatID = groups.getByProtocolID(groupId).getId();
        mainWindow.addMessage(new ChatMessageBlueprint(Chat.chatMessageType.FROM,
                contacts.getByIP(packet.getAddress()).getUsername(), message,
                "", contacts.getByIP(packet.getAddress()).getColor()), correspondingChatID);
    }

    public boolean isUsedID(int id) {
        return (contacts.getByID(id) != null || groups.getByID(id) != null);
    }

    @NotNull
    @Override
    public InetAddress[] getIpsFromGroup(int groupId) {
        return new InetAddress[0];
    }
}

File: src/logic/CallbackListener.kt
package logic

import gui.Chat
import gui.ChatMessageBlueprint
import gui.MainWindow
import protocol.ProtocolCallback
import protocol.getMacAddress
import java.net.DatagramPacket
import java.net.InetAddress
import java.util.*

class CallbackListener(private val mainWindow: MainWindow,
                       private val contacts: Contacts,
                       private val groups: Groups) : ProtocolCallback {
    var currentChatId = -1

    override fun hello(packet: DatagramPacket, username: String) {
        val contact = contacts.getOrCreateContact(getMacAddress(packet)!!,
                packet.address, username, UserColors.randomColor)
        // debug
        mainWindow.addContact(contact.username, "is now online", contact.color, contact.id)
        mainWindow.addNewChatById(contact.id)

        if (!contact.messages.isEmpty())
            mainWindow.addMessages(contact.messages.toTypedArray(), contact.id)
    }

    override fun world(packet: DatagramPacket, username: String) {
        val contact = contacts.getOrCreateContact(getMacAddress(packet)!!,
                packet.address, username, UserColors.randomColor)

        // debug
        mainWindow.addNewChatById(contact.id)
        mainWindow.addContact(contact.username, "is now online", contact.color, contact.id)

        if (!contact.messages.isEmpty())
            mainWindow.addMessages(contact.messages.toTypedArray(), contact.id)
    }

    override fun goodbye(packet: DatagramPacket) {
        val contact = contacts.getByMacAddress(packet) ?: return

        mainWindow.addMessage(ChatMessageBlueprint(Chat.chatMessageType.INFO, "ignored parameter",
                "${contact.username} has left the room.", "",
                contact.color), contact.id)
        mainWindow.setLastMessageText("is now offline", contact.id)
    }

    override fun typing(packet: DatagramPacket, typing: Boolean) {
        contacts.getByMacAddress(packet)?.id!!
                .let { id ->
                    if (typing) {
                        mainWindow.setContactWriting(id)
                    } else {
                        mainWindow.removeContactWriting(id)
                    }
                }
    }

    override fun message(packet: DatagramPacket, message: String) {
        val contact = contacts.getByMacAddress(packet)
        if (contact == null) {
            println("Message from unknown contact received: " + packet.address)
        } else {
            val chatMessageBlueprint = ChatMessageBlueprint(Chat.chatMessageType.FROM,
                    contact.username,
                    message, "", contact.color)

            mainWindow.addMessage(chatMessageBlueprint, contact.id)
            mainWindow.setLastMessageText(message, contact.id)

            contact.messages.add(chatMessageBlueprint)
        }
    }

    override fun existsGroupWithId(packet: DatagramPacket, id: Int): Boolean {
        // TODO: 19.12.2017 most likely should also return false, when a contact already uses this contactId
        // UPDATE: Nein!
        return groups.getByProtocolID(id) != null
    }

    override fun createGroup(packet: DatagramPacket, id: Int, members: Array<InetAddress>) {
        val membersAsContacts = ArrayList<Contact>()
        for (memberAddress in members) {
            val contact = contacts.getByMacAddress(getMacAddress(memberAddress)) ?: throw RuntimeException()

            membersAsContacts.add(contact)
        }
        groups.createGroup(id, membersAsContacts)
    }

    override fun denyGroup(packet: DatagramPacket) {
        // TODO: 15.12.2017 No clue what this means, or which steps should be taken.
    }

    override fun groupMessage(packet: DatagramPacket, groupId: Int, message: String) {
        // TODO: 19.12.2017 see todo
        val correspondingChatID = groups.getByProtocolID(groupId)!!.id
        mainWindow.addMessage(ChatMessageBlueprint(Chat.chatMessageType.FROM,
                contacts.getByMacAddress(packet)!!.username, message,
                "", contacts.getByMacAddress(packet)!!.color), correspondingChatID)
    }

    fun isUsedID(id: Int) = contacts.getById(id) != null

    override fun getIpsFromGroup(groupId: Int)
            = groups.getById(groupId)?.members!!.map { it.ip }.toTypedArray()

    override fun groupCreated(randId: Int, others: Array<out InetAddress>) {
        TODO("not implemented") //To change body of created functions use File | Settings | File Templates.
    }
}
